nucliadb_utils/nucliadb_utils/nats/demux.py

```diff
@@ -0,0 +1,310 @@
+
+import asyncio
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from functools import partial, total_ordering
+from typing import Any, Awaitable, Callable, Optional, Type
+
+from nats.aio.client import Msg
+
+from nucliadb_telemetry.errors import capture_exception
+from nucliadb_utils.nats.message_progress_updater import MessageProgressUpdater
+from nucliadb_utils.settings import nats_consumer_settings
+from nucliadb_utils.signals import Signal
+
+logger = logging.getLogger(__name__)
+
+
+Split = str
+DemuxProcessorWork = Any
+
+
+class DemuxProcessor(ABC):
+    """Interface for NatsDemultiplexer internal processing."""
+
+    @abstractmethod
+    def splitter(self, msg: Msg) -> tuple[Split, DemuxProcessorWork]:
+        """Decide in which split this message should go and return an appropiate
+        work unit from it.
+
+        NOTE: if work units will be added in an ordered data structure, the
+        return value should be sortable in the desired order.
+
+        """
+        ...
+
+    @abstractmethod
+    async def process(self, work: DemuxProcessorWork) -> bool:
+        """Process a work unit and return a boolean indicating processing
+        success or failure.
+
+        """
+        ...
+
+
+@total_ordering
+@dataclass
+class WorkUnit:
+    work_id: str
+    work: DemuxProcessorWork
+
+    def __eq__(self, other):
+        # A work unit is the same if we assign them the same id
+        if not isinstance(other, WorkUnit):
+            return NotImplemented
+        return self.work_id == other.work_id
+
+    def __lt__(self, other):
+        # Rely on DemuxProcessorWork implementing partial ordering
+        if not isinstance(other, WorkUnit):
+            return NotImplemented
+        return self.work.__lt__(other.work)
+
+    def __hash__(self):
+        return self.work_id.__hash__()
+
+
+@dataclass
+class WorkOutcome:
+    work_id: str
+    success: Optional[bool] = None
+    error: Optional[Exception] = None
+
+    def __post_init__(self):
+        if self.success is not None and self.error is not None:
+            raise ValueError("Can't create a work outcome with success and error")
+        elif self.success is None and self.error is None:
+            raise ValueError("Can't create a work outcome without success or error")
+
+
+work_unit_done = Signal(payload_model=WorkOutcome)
+
+
+class NatsDemultiplexer:
+    """Given a NATS consumer, allow concurrent message processing.
+
+    NatsDemultiplexer relies on a DemuxProcessor. It's `splitter` function will
+    be used to decide the split and the resulting work unit will be added into a
+    queue.
+
+    Concurrent workers will get work from their respective queues and use
+    `DemuxProcessor.process` to process it. Depending on the result, the message
+    will be acked or nacked to NATS.
+
+    If an exception is raised during any of this process, the message will be
+    automatically nacked.
```

Comment:
So it is responsibility of the specific worker to handle any unrecoverable exceptions and then ack, right?